File: ruby-sdk/semantria/session.rb
# encoding: utf-8
$LOAD_PATH << File.dirname(__FILE__) unless $LOAD_PATH.include?(File.dirname(__FILE__))

require 'authrequest'
require 'jsonserializer'
require 'xmlserializer'
require 'xmlhandlers'

class Session
   # Create a new instance
  def initialize(consumerKey, consumerSecret, serializer = nil, applicationName = nil)
    @host = 'https://api21.semantria.com'
	@wrapperName = 'Ruby'
    @consumerKey = consumerKey
    @consumerSecret = consumerSecret
	
	unless applicationName.nil?
		@applicationName = '%s.%s' % [applicationName, @wrapperName]
	else
		@applicationName = @wrapperName
	end
	
    unless serializer.nil?
      @serializer = serializer
      @format = @serializer.gettype()
    else
      #set default json serializer
      @serializer = JsonSerializer.new()
      @format = @serializer.gettype()
    end
  end
  
  def host= host
    @host = host
  end
  
  def host
    @host
  end
 
  def registerSerializer(serializer)
    unless serializer.nil?
      @serializer = serializer
      @format = serializer.gettype()
    else
      raise 'Parameter not found: %s' % serializer
    end
  end
  
  def setCallbackHandler(callback)
    if (callback.class < CallbackHandler)
      @callback = callback
    else
      raise 'Parameter is not subclass of CallbackHandler %s' % callback
    end
  end
  
  def createUpdateProxy()
    return {"added"=>[], "removed"=>[], "cloned"=>[]}
  end

  def getStatus()
    url = '%s/status.%s' % [@host, @format]
    return runRequest("GET", url, "get_status")
  end
 
  def verifySubscription()
    url = '%s/subscription.%s' % [@host, @format]
    return runRequest("GET", url, "get_subscription")
  end
 
  def getConfigurations()
    url = '%s/configurations.%s' % [@host, @format]
    result = runRequest("GET", url, "get_configurations")
    
    if result.nil?
      result = []
    end
    return result
  end
  
  def updateConfigurations(updates)
    url = '%s/configurations.%s' % [@host, @format]
    wrapper = getTypeWrapper("update_configurations")
    data = @serializer.serialize(updates, wrapper)
    return runRequest("POST", url, nil, data)
  end
  
  def getBlacklist(configId = nil)
    unless configId.nil?
      url = '%s/blacklist.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/blacklist.%s' % [@host, @format]
    end
    
    result = runRequest("GET", url, "get_blacklist")
    if result.nil?
      result = []
    end
    return result
  end
  
  def updateBlacklist(updates, configId = nil)
    unless configId.nil?
      url = '%s/blacklist.%s?config_id=%s' % [@host, @format, configId]
    else
       url = '%s/blacklist.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("update_blacklist")
    data = @serializer.serialize(updates, wrapper)
    return runRequest("POST", url, nil, data)
  end
  
  def getCategories(configId = nil)
    unless configId.nil?
      url = '%s/categories.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/categories.%s' % [@host, @format]
    end
    
    result = runRequest("GET", url, "get_categories")
    if result.nil?
      result = []
    end
    return result
  end
  
  def updateCategories(updates, configId = nil)
    unless configId.nil?
      url = '%s/categories.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/categories.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("update_categories")
    data = @serializer.serialize(updates, wrapper)
    return runRequest("POST", url, nil, data)
  end
  
  def getQueries(configId = nil)
    unless configId.nil?
      url = '%s/queries.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/queries.%s' % [@host, @format]
    end
    
    result = runRequest("GET", url, "get_queries")
    if result.nil?
      result = []
    end
    return result
  end
  
  def updateQueries(updates, configId = nil)
    unless configId.nil?
      url = '%s/queries.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/queries.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("update_queries")
    data = @serializer.serialize(updates, wrapper)
    return runRequest("POST", url, nil, data)
  end
  
  def getSentimentPhrases(configId = nil)
    unless configId.nil?
      url = '%s/sentiment.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/sentiment.%s' % [@host, @format]
    end
    
    result = runRequest("GET", url, "get_sentiment_phrases")
    if result.nil?
      result = []
    end
    return result
  end
  
  def updateSentimentPhrases(updates, configId = nil)
    unless configId.nil?
      url = '%s/sentiment.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/sentiment.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("update_sentiment_phrases")
    data = @serializer.serialize(updates, wrapper)
    return runRequest("POST", url, nil, data)
  end  
    
  def getEntities(configId = nil)
    unless configId.nil?
      url = '%s/entities.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/entities.%s' % [@host, @format]
    end
    
    result = runRequest("GET", url, "get_entities")
    if result.nil?
      result = []
    end
    return result
  end
  
  def updateEntities(updates, configId = nil)
    unless configId.nil?
      url = '%s/entities.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/entities.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("update_entities")
    data = @serializer.serialize(updates, wrapper)
    return runRequest("POST", url, nil, data)
  end
  
  def queueDocument(task, configId = nil)
    unless configId.nil?
      url = '%s/document.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/document.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("queue_document")
    data = @serializer.serialize(task, wrapper)
    result = runRequest("POST", url, "get_processed_documents", data)
    if (result != nil && result.is_a?(Array))
      onDocsAutoResponse(result)
      return 200
    else 
      return result
    end
  end
  
  def queueBatchOfDocuments(batch, configId = nil)
    unless configId.nil?
      url = '%s/document/batch.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/document/batch.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("queue_batch_documents")
    data = @serializer.serialize(batch, wrapper)
    result = runRequest("POST", url, "get_processed_documents", data)
    if (result != nil && result.is_a?(Array))
      onDocsAutoResponse(result)
      return 200
    else 
      return result
    end
  end

  def getDocument(id, configId = nil)
    unless configId.nil?
      unless id.nil?
        url = '%s/document/%s.%s?config_id=%s' % [@host, id, @format, configId]
      else
        raise 'Parameter is nil %s' % id
      end
    else
      url = '%s/document/%s.%s' % [@host, id, @format]
    end
    
    return runRequest("GET", url, "get_document")
  end

  def cancelDocument(id, configId = nil)
    unless configId.nil?
      unless id.nil?
        url = '%s/document/%s.%s?config_id=%s' % [@host, id, @format, configId]
      else
        raise 'Parameter is nil %s' % id
      end
    else
      url = '%s/document/%s.%s' % [@host, id, @format]
    end

    return runRequest("DELETE", url)
  end

  def getProcessedDocuments(configId = nil)
    unless configId.nil?
      url = '%s/document/processed.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/document/processed.%s' % [@host, @format]
    end
    
    result = runRequest("GET", url, "get_processed_documents")
    if result.nil?
      result = []
    end
    return result
  end
     
  def queueCollection(task, configId = nil)
    unless configId.nil?
      url = '%s/collection.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/collection.%s' % [@host, @format]
    end

    wrapper = getTypeWrapper("queue_collection")
    data = @serializer.serialize(task, wrapper)
    result = runRequest("POST", url, "get_processed_collections", data)
    if (result != nil && result.is_a?(Array))
      onCollsAutoResponse(result)
      return 200
    else 
      return result
    end
  end

  def getCollection(id, configId = nil)
    unless configId.nil?
      unless id.nil?
        url = '%s/collection/%s.%s?config_id=%s' % [@host, id, @format, configId]
      else
        raise 'Parameter is nil %s' % id
      end
    else
      url = '%s/collection/%s.%s' % [@host, id, @format]
    end
    
    return runRequest("GET", url, "get_collection")
  end

  def cancelCollection(id, configId = nil)
    unless configId.nil?
      unless id.nil?
        url = '%s/collection/%s.%s?config_id=%s' % [@host, id, @format, configId]
      else
        raise 'Parameter is nil %s' % id
      end
    else
      url = '%s/collection/%s.%s' % [@host, id, @format]
    end

    return runRequest("DELETE", url)
  end

  def getProcessedCollections(configId = nil)
    unless configId.nil?
      url = '%s/collection/processed.%s?config_id=%s' % [@host, @format, configId]
    else
      url = '%s/collection/processed.%s' % [@host, @format]
    end
    
    result = runRequest("GET", url, "get_processed_collections")
    if result.nil?
      result = []
    end
    return result
  end
   
  private
  def runRequest(method, url, type = nil, postData = nil)
    request = AuthRequest.new(@consumerKey, @consumerSecret, @applicationName)
    onRequest({"method"=>method, "url"=>url, "message"=>postData}) 
    response = request.authWebRequest(method, url, postData)
    onResponse({"status"=>response["status"], "reason"=>response["reason"], "message"=>response["data"]}) 
    
    status = response["status"]
    message = response["reason"]
    
    if (response["data"] != "")
      message = response["data"]
    end
  
    if (method == "DELETE")
      if (status == 200 or status == 202)
        return status
      else
        resolveError(status, message)
        return status
      end
    else
      if (status == 200)
        handler = getTypeHandler(type) 
        message = @serializer.deserialize(response["data"], handler)
        return message
      elsif (status == 202)
        if (method == "POST")
          return status
        else
          return nil
        end
      else
        resolveError(status, message)
      end
    end
  end
        
  def getTypeHandler(type)
    if @serializer.gettype() == "json"
        return nil
    end

    #only for xml serializer
    if type == "get_status"
        return GetStatusHandler.new()
    elsif type == "get_subscription"
        return GetSubscriptionHandler.new()
    elsif type == "get_configurations"
        return GetConfigurationsHandler.new()
    elsif type == "get_blacklist"
        return GetBlacklistHandler.new()
    elsif type == "get_categories"
        return GetCategoriesHandler.new()
    elsif type == "get_queries"
        return GetQueriesHandler.new()
    elsif type == "get_sentiment_phrases"
        return GetSentimentPhrasesHandler.new()	
    elsif type == "get_entities"
        return GetEntitiesHandler.new()
    elsif type == "get_document"
        return GetDocumentHandler.new()
    elsif type == "get_processed_documents"
        return GetProcessedDocumentsHandler.new()
    elsif type == "get_collection"
        return GetCollectionHandler.new()
    elsif type == "get_processed_collections"
        return GetProcessedCollectionsHandler.new()
    else
        return nil
    end
  end
 
  def getTypeWrapper(type)
    if @serializer.gettype() == "json"
      return nil
    end

    #only for xml serializer
    if type == "update_configurations"
      return {"root"=>"configurations", "added"=>"configuration", "removed"=>"configuration"}
    elsif type == "update_blacklist"
      return {"root"=>"blacklist", "added"=>"item", "removed"=>"item"}
    elsif type == "update_categories"
      return {"root"=>"categories", "added"=>"category", "removed"=>"category", "samples"=>"sample"}
    elsif type == "update_queries"
      return {"root"=>"queries", "added"=>"query", "removed"=>"query"}
    elsif type == "update_sentiment_phrases"
      return {"root"=>"phrases", "added"=>"phrase", "removed"=>"phrase"}      
    elsif type == "update_entities"
      return {"root"=>"entities", "added"=>"entity", "removed"=>"entity"}
    elsif type == "queue_document"
      return {"root"=>"document"}
    elsif type == "queue_batch_documents"
      return {"root"=>"documents", "item"=>"document"}
    elsif type == "queue_collection"
      return {"root"=>"collection", "documents"=>"document"}
    else
      return nil
    end
  end
  
  def resolveError(status, message = nil)
    if (status == 400 || status == 401 || status == 402 || status == 403 || status == 406 || status == 500)
      onError({"status"=>status, "message"=>message})
    else
      raise 'HTTP error was found with status: %s and message: %s' % [status, message]
    end
  end
  
  def onRequest(request) 
    unless @callback.nil?
      @callback.onRequest(self, request) 
    end
  end 
  
  def onResponse(response) 
    unless @callback.nil? 
      @callback.onResponse(self, response) 
    end
  end
  
  def onError(response) 
    unless @callback.nil?
      @callback.onError(self, response) 
    end 
  end
  
  def onDocsAutoResponse(response)
    unless @callback.nil? 
      @callback.onDocsAutoResponse(self, response) 
    end 
  end
    
  def onCollsAutoResponse(response)
    unless @callback.nil? 
      @callback.onCollsAutoResponse(self, response) 
    end 
  end
end 

class CallbackHandler
  def initialize
    raise "Don't instantiate me!" if abstract_class?
    puts "Instantiated!"
  end
  
  private
  def abstract_class?
    self.class == CallbackHandler
  end

  def onRequest(sender, args)
    raise NoMethodError.new("Abstract method onRequest")
  end
  
  def onResponse(sender, args)
    raise NoMethodError.new("Abstract method onResponse")
   end
  
  def onError(sender, args)
    raise NoMethodError.new("Abstract method onError")
  end
  
  def onDocsAutoResponse(sender, args)
    raise NoMethodError.new("Abstract method onDocsAutoResponse")
  end
    
  def onCollsAutoResponse(sender, args)
    raise NoMethodError.new("Abstract method onCollsAutoResponse")
  end
end

File: php-sdk/Semantria/Session.php
<?php

require_once 'xmlhandlers.php';
require_once 'common.php';

class Semantria_Session
{
    // the consumer key and secret
    protected $consumerKey;
    protected $consumerSecret;
    protected $applicationName;

    protected $serializer;
    protected $format = 'json';
    protected $wrapperName = 'PHP';

    public $host = "https://api21.semantria.com";

    public function __construct($consumerKey = null, $consumerSecret = null, $serializer = null, $applicationName = null)
    {
        if (empty($consumerKey)) {
            throw new Exception('Parameter is null or empty "'.$consumerKey.'"');
        }
        if (empty($consumerSecret)) {
            throw new Exception('Parameter is null or empty "'.$consumerSecret.'"');
        }

        $this->consumerKey = $consumerKey;
        $this->consumerSecret = $consumerSecret;

        if (empty($applicationName)) {
            $this->applicationName = $this->wrapperName;
        } else {
            $this->applicationName = $applicationName . '.' . $this->wrapperName;
        }

        if (empty($serializer)) {
            $serializer = new Semantria_JsonSerializer();
        }

        $this->serializer = $serializer;
        $this->format = $serializer->getType();
    }

    public function setCallbackHandler(Semantria_CallbackHandler_Default &$callback)
    {
        $this->callback = $callback;
    }

    public function registerSerializer($serializer)
    {
        if (!empty($serializer)) {
            $this->serializer = $serializer;
            $this->format = $serializer->getType();
        } else {
            throw new Exception('Parameter is null or empty "'.$serializer.'"');
        }
    }

    public function createUpdateProxy()
    {
        return array("added"=>array(), "removed"=>array(), "cloned"=>array());
    }

    public function getStatus()
    {
        $url = string_format('{0}/status.{1}', $this->host, $this->format);
        return $this->runRequest("GET", $url, "get_status");
    }

    public function verifySubscription()
    {
        $url = string_format('{0}/subscription.{1}', $this->host, $this->format);
        return $this->runRequest("GET", $url, "get_subscription");
    }

    public function getConfigurations()
    {
        $url = string_format('{0}/configurations.{1}', $this->host, $this->format);
        $result = $this->runRequest("GET", $url, "get_configurations");

        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    public function updateConfigurations($updates)
    {
        $url = string_format('{0}/configurations.{1}', $this->host, $this->format);
        $wrapper = $this->getTypeWrapper("update_configurations");
        $data = $this->serializer->serialize($updates, $wrapper);
        return $this->runRequest("POST", $url, null, $data);
    }

    public function getBlacklist($configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/blacklist.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/blacklist.{1}', $this->host, $this->format);
        }

        $result = $this->runRequest("GET", $url, "get_blacklist");
        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    public function updateBlacklist($updates, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/blacklist.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/blacklist.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("update_blacklist");
        $data = $this->serializer->serialize($updates, $wrapper);
        return $this->runRequest("POST", $url, null, $data);
    }

    public function getCategories($configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/categories.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/categories.{1}', $this->host, $this->format);
        }

        $result = $this->runRequest("GET", $url, "get_categories");
        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    public function updateCategories($updates, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/categories.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/categories.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("update_categories");
        $data = $this->serializer->serialize($updates, $wrapper);
        return $this->runRequest("POST", $url, null, $data);
    }

    public function getQueries($configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/queries.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/queries.{1}', $this->host, $this->format);
        }

        $result = $this->runRequest("GET", $url, "get_queries");
        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    public function updateQueries($updates, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/queries.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/queries.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("update_queries");
        $data = $this->serializer->serialize($updates, $wrapper);
        return $this->runRequest("POST", $url, null, $data);
    }

    public function getSentimentPhrases($configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/sentiment.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/sentiment.{1}', $this->host, $this->format);
        }

        $result = $this->runRequest("GET", $url, "get_sentiment_phrases");
        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    public function updateSentimentPhrases($updates, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/sentiment.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/sentiment.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("update_sentiment_phrases");
        $data = $this->serializer->serialize($updates, $wrapper);
        return $this->runRequest("POST", $url, null, $data);
    }

    public function getEntities($configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/entities.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/entities.{1}', $this->host, $this->format);
        }

        $result = $this->runRequest("GET", $url, "get_entities");
        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    public function updateEntities($updates, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/entities.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/entities.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("update_entities");
        $data = $this->serializer->serialize($updates, $wrapper);
        return $this->runRequest("POST", $url, null, $data);
    }

    public function queueDocument($task, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/document.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/document.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("queue_document");
        $data = $this->serializer->serialize($task, $wrapper);

        $result = $this->runRequest("POST", $url, "get_processed_documents", $data);
        if (isset($result) && !is_int($result)) {
            $this->onDocsAutoResponse($result);
            return 200;
        } else {
            return $result;
        }
    }

    public function queueBatchOfDocuments($batch, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/document/batch.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/document/batch.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("queue_batch_documents");
        $data = $this->serializer->serialize($batch, $wrapper);
        $result = $this->runRequest("POST", $url, "get_processed_documents", $data);
        if (isset($result) && !is_int($result)) {
            $this->onDocsAutoResponse($result);
            return 200;
        } else {
            return $result;
        }
    }

    public function getDocument($id, $configId=null)
    {
        if (isset($configId)) {
            if (isset($id)) {
                $url = string_format('{0}/document/{1}.{2}?config_id={3}', $this->host, $id, $this->format, $configId);
            } else {
                throw new Exception('Parameter is null or empty "'.$id.'"');
            }

        } else {
            $url = string_format('{0}/document/{1}.{2}', $this->host, $id, $this->format);
        }

        return $this->runRequest("GET", $url, "get_document");
    }

    public function cancelDocument($id, $configId=null)
    {
        if (isset($configId)) {
            if (isset($id)) {
                $url = string_format('{0}/document/{1}.{2}?config_id={3}', $this->host, $id, $this->format, $configId);
            } else {
                throw new Exception('Parameter is null or empty "'.$id.'"');
            }

        } else {
            $url = string_format('{0}/document/{1}.{2}', $this->host, $id, $this->format);
        }

        return $this->runRequest("DELETE", $url);
    }

    public function getProcessedDocuments($configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/document/processed.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/document/processed.{1}', $this->host, $this->format);
        }

        $result = $this->runRequest("GET", $url, "get_processed_documents");
        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    public function queueCollection($task, $configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/collection.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/collection.{1}', $this->host, $this->format);
        }

        $wrapper = $this->getTypeWrapper("queue_collection");
        $data = $this->serializer->serialize($task, $wrapper);
        $result = $this->runRequest("POST", $url, "get_processed_collections", $data);
        if (isset($result) && !is_int($result)) {
            $this->onCollsAutoResponse($result);
            return 200;
        } else {
            return $result;
        }
    }

    public function getCollection($id, $configId=null)
    {
        if (isset($configId)) {
            if (isset($id)) {
                $url = string_format('{0}/collection/{1}.{2}?config_id={3}', $this->host, $id, $this->format, $configId);
            } else {
                throw new Exception('Parameter is null or empty "'.$id.'"');
            }

        } else {
            $url = string_format('{0}/collection/{1}.{2}', $this->host, $id, $this->format);
        }

        return $this->runRequest("GET", $url, "get_collection");
    }

    public function cancelCollection($id, $configId=null)
    {
        if (isset($configId)) {
            if (isset($id)) {
                $url = string_format('{0}/collection/{1}.{2}?config_id={3}', $this->host, $id, $this->format, $configId);
            } else {
                throw new Exception('Parameter is null or empty "'.$id.'"');
            }

        } else {
            $url = string_format('{0}/collection/{1}.{2}', $this->host, $id, $this->format);
        }

        return $this->runRequest("DELETE", $url);
    }

    public function getProcessedCollections($configId=null)
    {
        if (isset($configId)) {
            $url = string_format('{0}/collection/processed.{1}?config_id={2}', $this->host, $this->format, $configId);
        } else {
            $url = string_format('{0}/collection/processed.{1}', $this->host, $this->format);
        }

        $result = $this->runRequest("GET", $url, "get_processed_collections");
        if (!isset($result)) {
            $result = array();
        }
        return $result;
    }

    private function runRequest($method, $url, $type=null, $postData = null)
    {
        $this->onRequest(array("method"=>$method, "url"=>$url, "message"=>$postData));
        $request = new Semantria_AuthRequest($this->consumerKey, $this->consumerSecret);

        $response = $request->authWebRequest($method, $url, $postData);
        $this->onResponse($response);

        $status = $response["status"];
        $message = $response["message"];

        if ($method == "DELETE") {
            if ($status == 200 or $status == 202) {
                return $status;
            } else {
                $this->resolveError($status, $message);
                return $status;
            }
        } else {
            if ($status == 200) {
                $handler = $this->getTypeHandler($type);
                $obj = $this->serializer->deserialize($message, $handler);
                return $obj;
            } elseif ($status == 202) {
                if ($method == "POST") {
                    return $status;
                } else {
                    return null;
                }
            } else {
                $this->resolveError($status, $message);
            }
        }
    }

    private function resolveError($status, $message=null)
    {
        if ($status == 400 || $status == 401 || $status == 402 || $status == 403 || $status == 406 || $status == 500)
            $this->onError(array("status"=>$status, "message"=>$message));
        else
            throw new Exception('Exception code: '.$status.' and message: '.$message);
    }

    private function getTypeHandler($type)
    {
        if ($this->serializer->gettype() == "json") {
            return null;
        }

        #only for xml serializer
        if ($type == "get_status") {
            return new Semantria_XmlHandler_Status();
        } elseif ($type == "get_subscription") {
            return new GetSubscriptionHandler();
        } elseif ($type == "get_configurations") {
            return new GetConfigurationsHandler();
        } elseif ($type == "get_blacklist") {
            return new GetBlacklistHandler();
        } elseif ($type == "get_categories") {
            return new GetCategoriesHandler();
        } elseif ($type == "get_queries") {
            return new GetQueriesHandler();
        } elseif ($type == "get_sentiment_phrases") {
            return new GetSentimentPhrasesHandler();
        } elseif ($type == "get_entities") {
            return new GetEntitiesHandler();
        } elseif ($type == "get_document") {
            return new GetDocumentHandler();
        } elseif ($type == "get_processed_documents") {
            return new GetProcessedDocumentsHandler();
        } elseif ($type == "get_collection") {
            return new GetCollectionHandler();
        } elseif ($type == "get_processed_collections") {
            return new GetProcessedCollectionsHandler();
        } else {
            return null;
        }
    }

    private function getTypeWrapper($type)
    {
        if ($this->serializer->gettype() == "json") {
            return null;
        }

        #only for xml serializer
        if ($type == "update_configurations") {
            return array("root"=>"configurations", "added"=>"configuration", "removed"=>"configuration");
        } elseif ($type == "update_blacklist") {
            return array("root"=>"blacklist", "added"=>"item", "removed"=>"item");
        } elseif ($type == "update_categories") {
            return array("root"=>"categories", "added"=>"category", "removed"=>"category", "samples"=>"sample");
        } elseif ($type == "update_queries") {
            return array("root"=>"queries", "added"=>"query", "removed"=>"query");
        } elseif ($type == "update_sentiment_phrases") {
            return array("root"=>"phrases", "added"=>"phrase", "removed"=>"phrase");
        } elseif ($type == "update_entities") {
            return array("root"=>"entities", "added"=>"entity", "removed"=>"entity");
        } elseif ($type == "queue_document") {
            return array("root"=>"document");
        } elseif ($type == "queue_batch_documents") {
            return array("root"=>"documents", "item"=>"document");
        } elseif ($type == "queue_collection") {
            return array("root"=>"collection", "documents"=>"document");
        } else {
            return null;
        }
    }

    private function onRequest($request)
    {
        if (isset($this->callback)) {
            call_user_func_array(array($this->callback, "onRequest"), array(&$this, $request));
        }
    }

    private function onResponse($response)
    {
        if (isset($this->callback)) {
            call_user_func_array(array($this->callback, "onResponse"), array(&$this, $response));
        }
    }

    private function onError($response)
    {
        if (isset($this->callback)) {
            call_user_func_array(array($this->callback, "OnError"), array(&$this, $response));
        }
    }

    private function onDocsAutoResponse($response)
    {
        if (isset($this->callback)) {
            call_user_func_array(array($this->callback, "OnDocsAutoResponse"), array(&$this, $response));
        }
    }

    private function onCollsAutoResponse($response)
    {
        if (isset($this->callback)) {
            call_user_func_array(array($this->callback, "OnCollsAutoResponse"), array(&$this, $response));
        }
    }
}

